core/arch/x86/decode_fast.c

```diff
@@ -537,24 +538,38 @@ decode_sizeof(dcontext_t *dcontext, byte *start_pc,
                 sz += 1;
                 /* up to caller to check for addr prefix! */
                 break;
+            case EVEX_PREFIX_OPCODE: {
+                /* If 64-bit mode or EVEX.R' bit is flipped, this is evex */
+                if (X64_MODE_DC(dcontext) || TEST(0x10, *(pc + 1))) {
+                    evex_prefix = true;
+                }
+                /* Fall-through is deliberate, EVEX is handled through VEX below */
+            }
             case VEX_3BYTE_PREFIX_OPCODE:
             case VEX_2BYTE_PREFIX_OPCODE: {
                 /* If 64-bit mode or mod selects for register, this is vex */
-                if (X64_MODE_DC(dcontext) || TESTALL(MODRM_BYTE(3, 0, 0), *(pc + 1))) {
+                if (evex_prefix || X64_MODE_DC(dcontext) ||
+                    TESTALL(MODRM_BYTE(3, 0, 0), *(pc + 1))) {
                     /* Assumptions:
                      * - no vex-encoded instr size differs based on vex.w,
                      *   so we don't bother to set qword_operands
                      * - no vex-encoded instr size differs based on prefixes,
                      *   so we don't bother to decode vex.pp
                      */
-                    bool vex3 = (opc == 0xc4);
+                    bool vex3 = (opc == VEX_3BYTE_PREFIX_OPCODE);
                     byte vex_mm = 0;
                     opc = (uint) * (++pc); /* 2nd vex prefix byte */
                     sz += 1;
                     if (vex3) {
                         vex_mm = (byte)(opc & 0x1f);
                         opc = (uint) * (++pc); /* 3rd vex prefix byte */
                         sz += 1;
+                    } else if (evex_prefix) {
```

Comment:
I would add a `XXX i#1312:...` comment about needing to update the sizeof_* tables.  (Otherwise it is confusing for the reader to see this evex support and likely assume avx512 is fully supported for this function.)